search.go

```diff
@@ -75,33 +75,67 @@ func Find(index ColumnIndex, value Value, cmp func(Value, Value) int) int {
 
 func binarySearch(index ColumnIndex, value Value, cmp func(Value, Value) int) int {
 	n := index.NumPages()
-	i := 0
-	j := n
+	curIdx := 0
+	topIdx := n
 
-	for (j - i) > 1 {
-		k := ((j - i) / 2) + i
-		c := cmp(value, index.MinValue(k))
+	// while there's at least one more page to check
+	for (topIdx - curIdx) > 1 {
+
+		// nextIdx is halfway between curIdx and topIdx
+		nextIdx := ((topIdx - curIdx) / 2) + curIdx
+
+		smallerThanMin := cmp(value, index.MinValue(nextIdx))
+		greaterThanMax := cmp(value, index.MaxValue(nextIdx))
 
 		switch {
-		case c < 0:
-			j = k
-		case c > 0:
-			i = k
+		// search below this page
+		case smallerThanMin < 0:
+			topIdx = nextIdx
+		// search above this page
+		case greaterThanMax > 0:
+			curIdx = nextIdx
+		case smallerThanMin == 0:
+			// this case is hit when winValue == value of nextIdx
+			// we must check below this index to find if there's
+			// another page before this.
+			// e.g. searching for first page 3 is in:
+			// [1,2,3]
+			// [3,4,5]
+			// [6,7,8]
+
+			// if the page proceeding this has a maxValue matching the value we're
+			// searching, continue the search.
+			// otherwise, we can return early
+			//
+			// cases covered by else block
+			// if cmp(value, index.MaxValue(nextIdx-1)) < 0: the value is only in this page
+			// if cmp(value, index.MaxValue(nextIdx-1)) > 0: we've got a sorting problem with overlapping pages
+			if nextIdx-1 > curIdx && cmp(value, index.MaxValue(nextIdx-1)) == 0 {
+				topIdx = nextIdx
```

Comment:
It maybe more efficient to do a linear search from this on out, but I don't want to assume that outright (e.g. null values at the beginning of a column could have many pages worth of data, so still worth binary searching it?).

Reply:
I don't think we can assume that nulls will be at the beginning, they might be placed at the end depending on how the column is configured.

Reply:
Sure, it was just an example argument against doing a linear search in reverse.